sklearn/linear_model/tests/test_sgd.py

```diff
@@ -113,1102 +140,1181 @@ def decision_function(self, X, *args, **kw):
 ###############################################################################
 # Tests common to classification and regression
 
-class CommonTest:
-
-    def factory(self, **kwargs):
-        if "random_state" not in kwargs:
-            kwargs["random_state"] = 42
+# a simple implementation of ASGD to use for testing
+# uses squared loss to find the gradient
+def asgd(klass, X, y, eta, alpha, weight_init=None, intercept_init=0.0):
+    if weight_init is None:
+        weights = np.zeros(X.shape[1])
+    else:
+        weights = weight_init
+
+    average_weights = np.zeros(X.shape[1])
+    intercept = intercept_init
+    average_intercept = 0.0
+    decay = 1.0
 
-        if "tol" not in kwargs:
-            kwargs["tol"] = None
-        if "max_iter" not in kwargs:
-            kwargs["max_iter"] = 5
+    # sparse data has a fixed decay of .01
+    if klass in (SparseSGDClassifier, SparseSGDRegressor):
+        decay = .01
 
-        return self.factory_class(**kwargs)
+    for i, entry in enumerate(X):
+        p = np.dot(entry, weights)
+        p += intercept
+        gradient = p - y[i]
+        weights *= 1.0 - (eta * alpha)
+        weights += -(eta * gradient * entry)
+        intercept += -(eta * gradient) * decay
+
+        average_weights *= i
+        average_weights += weights
+        average_weights /= i + 1.0
+
+        average_intercept *= i
+        average_intercept += intercept
+        average_intercept /= i + 1.0
+
+    return average_weights, average_intercept
+
+
+def _test_warm_start(klass, X, Y, lr):
+    # Test that explicit warm restart...
+    clf = klass(alpha=0.01, eta0=0.01, shuffle=False,
+                learning_rate=lr)
+    clf.fit(X, Y)
+
+    clf2 = klass(alpha=0.001, eta0=0.01, shuffle=False,
+                 learning_rate=lr)
+    clf2.fit(X, Y,
+             coef_init=clf.coef_.copy(),
+             intercept_init=clf.intercept_.copy())
+
+    # ... and implicit warm restart are equivalent.
+    clf3 = klass(alpha=0.01, eta0=0.01, shuffle=False,
+                 warm_start=True, learning_rate=lr)
+    clf3.fit(X, Y)
+
+    assert_equal(clf3.t_, clf.t_)
+    assert_array_almost_equal(clf3.coef_, clf.coef_)
+
+    clf3.set_params(alpha=0.001)
+    clf3.fit(X, Y)
+
+    assert_equal(clf3.t_, clf2.t_)
+    assert_array_almost_equal(clf3.coef_, clf2.coef_)
+
+
+@pytest.mark.parametrize('klass', [SGDClassifier, SparseSGDClassifier,
+                                   SGDRegressor, SparseSGDRegressor])
+@pytest.mark.parametrize('lr',
+                         ["constant", "optimal", "invscaling", "adaptive"])
+def test_warm_start(klass, lr):
+    _test_warm_start(klass, X, Y, lr)
```

Comment:
You can now just include the code above from this `_test_warm_start` here in the test function? (now it is only called once, before 4 times)

Reply:
_test_warm_start is called twice with another X2, Y2 below